types/babel__traverse/index.d.ts

```diff
@@ -219,74 +328,88 @@ export type VisitNodeFunction<S, P extends Node> = (this: S, path: NodePath<P>,
 }
 
+export type NodeKeyOfArrays<T extends Node> = {
+    [P in keyof T]-?: T[P] extends Array<Node | null | undefined> ? P : never;
+}[keyof T];
+
+export type NodeKeyOfNodes<T extends Node> = {
+    [P in keyof T]-?: T[P] extends Node | null | undefined ? P : never;
+}[keyof T];
+
 export type NodePaths<T extends Node | readonly Node[]> = T extends readonly Node[]
     ? { -readonly [K in keyof T]: NodePath<Extract<T[K], Node>> }
     : T extends Node
     ? [NodePath<T>]
     : never;
 
+type NodeListType<N, K extends keyof N> = N[K] extends Array<infer P> ? (P extends Node ? P : never) : never;
+
+type NodesInsertionParam<T extends Node> = T | readonly T[] | [T, ...T[]];
+
 export class NodePath<T = Node> {
-    constructor(hub: Hub, parent: Node);
+    constructor(hub: HubInterface, parent: Node);
     parent: Node;
     hub: Hub;
+    data: Record<string | symbol, unknown>;
+    context: TraversalContext;
+    scope: Scope;
     contexts: TraversalContext[];
-    data: object;
+    state: any;
+    opts: any; // exploded TraverseOptions
+    skipKeys: Record<string, boolean> | null;
+    parentPath: T extends t.Program ? null : NodePath;
+    container: Node | Node[] | null;
+    listKey: string | null;
+    key: string | number | null;
+    node: T;
+    type: T extends Node ? T['type'] : T extends null | undefined ? undefined : Node['type'] | undefined;
     shouldSkip: boolean;
     shouldStop: boolean;
     removed: boolean;
-    state: any;
-    opts: object;
-    skipKeys: object;
-    parentPath: T extends t.Program ? null : NodePath;
-    context: TraversalContext;
-    container: object | object[];
-    listKey: string;
     inList: boolean;
     parentKey: string;
-    key: string | number;
-    node: T;
-    scope: Scope;
-    type: T extends null | undefined ? undefined : T extends Node ? T['type'] : string | undefined;
     typeAnnotation: object;
 
+    static get<C extends Node, K extends keyof C>(opts: {
+        hub?: HubInterface;
+        parentPath: NodePath | null;
+        parent: Node;
+        container: C;
+        key: K;
+    }): NodePath<C[K]>;
+    static get<C extends Node, L extends NodeKeyOfArrays<C>>(opts: {
+        hub?: HubInterface;
+        parentPath: NodePath | null;
+        parent: Node;
+        container: C;
+        listKey: L;
+        key: number;
+    }): C[L] extends Array<Node | null | undefined> ? NodePath<C[L][number]> : never;
+
     getScope(scope: Scope): Scope;
 
-    setData(key: string, val: any): any;
+    setData(key: string | symbol, val: any): any;
 
-    getData(key: string, def?: any): any;
+    getData(key: string | symbol, def?: any): any;
 
-    hasNode(): this is NodePath<NonNullable<this['node']>>;
+    hasNode(): this is NodePath<Exclude<T, null | undefined>>;
 
-    buildCodeFrameError<TError extends Error>(msg: string, Error?: new (msg: string) => TError): TError;
+    buildCodeFrameError(msg: string, Error?: ErrorConstructor): Error;
```

Comment:
I think this function (and everything using `ErrorConstructor`, really) should be reverted, because when using typescript projects with `@types/node@20` present, the type of ErrorConstructor conflicts with that of the ErrorConstructor from lib@es5 (built in TS).

As is, folks calling `buildCodeFrameError` need cannot past vanilla errors (or errors extending from `Error`) because Error !== ErrorConstructor for all environments.